stonesoup/updater/particle.py

```diff
@@ -70,7 +70,236 @@ def predict_measurement(self, state_prediction, measurement_model=None,
+                particle, predictor.position_mapping, transition, predictor.transition_model, time_interval)
+
+            particle.model_probabilities = new_model_probabilities
+
+            hypothesis.measurement.state_vector = np.reshape(
+                hypothesis.measurement.state_vector, (-1, 1)
+            )
+
+            prob_y_given_x = measurement_model.pdf(
+                hypothesis.measurement, particle,
+                **kwargs)
+
+            particle.weight *= prob_y_given_x
+
+        # Normalise the weights
+        sum_w = Probability.sum(
+            i.weight for i in hypothesis.prediction.particles)
+        for particle in hypothesis.prediction.particles:
+            particle.weight /= sum_w
+
+        new_particles = self.resampler.resample(
+            hypothesis.prediction.particles)
+
+        return ParticleStateUpdate(new_particles,
+                                   hypothesis,
+                                   timestamp=hypothesis.measurement.timestamp)
+
+    @lru_cache()
+    def predict_measurement(self, state_prediction, measurement_model=None,
+                            **kwargs):
+
+        if measurement_model is None:
+            measurement_model = self.measurement_model
+
+        new_particles = []
+        for particle in state_prediction.particles:
+            new_state_vector = measurement_model.function(
+                particle.state_vector, noise=0, **kwargs)
+            new_particles.append(
+                RaoBlackwellisedParticle(new_state_vector,
+                                         weight=particle.weight,
+                                         parent=particle.parent,
+                                         model_probabilities=particle.model_probabilities))
+
+        return ParticleMeasurementPrediction(
+            new_particles, timestamp=state_prediction.timestamp)
+
+    @staticmethod
+    def calculate_model_probabilities(particle, position_mapping,
+                                      transition_matrix, model_list, time_interval):
+
+        """Calculates the new model probabilities based
+            on the ones calculated in the previous time step"""
+
+        previous_probabilities = particle.model_probabilities
+
+        denominator_components = []
+        # Loop over the current models m_k
+        for l in range(len(previous_probabilities)):
+            selected_model_sum = 0
+            # Loop over the previous models m_k-1
+            for i, model in enumerate(model_list):
+                # Looks up p(m_k|m_k-1)
+                # Note: if p(m_k|m_k-1) = 0 then p(m_k|x_1:k) = 0
+                transition_probability = transition_matrix[i][l]
+                if transition_probability == 0:
+                    break
+                # Getting required states to apply the model to that state vector
+                parent_required_state_space = particle.parent.state_vector[
```

Comment:
Should this be the parent particle? Or should it be the prior?

Reply:
I can see now that this is the prior, as the predictor assigned the parent to this. I think the parent variable should be reserved for resampling really.